### PR TITLE
Add OpenTelemetry OTLP Exporter

### PR DESCRIPTION
**Pull request by luneo7**
Adds OpenTelemetry OTLP Exporter and addresses features requests in https://github.com/quarkusio/quarkus/issues/17640

### CODE REVIEW

**Comment by kenfinnigan** [comment]
@luneo7 thanks for the contribution!

I have a few suggestions on the current changes:

1. Don't worry about moving the `LateBoundBatchSpanProcessor` to a common extension, just duplicate it within the two exporter extensions. The reason is I intend to propose the class to the OpenTelemetry Java project soon, so there should be a point where we don't need it at all in Quarkus code and we can use it from OpenTelemetry.
2. It would be preferred to separate the changes for adding an OTLP Exporter and modifying how the providers are configured into separate PRs to keep the changes isolated from one another. This helps with any potential backporting of a change to an older release
3. We need a slightly different approach for Propagators and Resources as there can be many of them as opposed to one. It was on my list to look at soon, so you can continue to work on a solution with my guidance, or I can look to add it, whatever you prefer.
4. In the `IdGenerator` case, if we were able to dispense with the `@DefaultBean` and only set it onto the provider builder if one was present that would be preferable as it saves on creating a duplicate object instance.

Let me know if you have any questions.

**Comment by luneo7** [comment]
Hi @kenfinnigan =]

I can do 1, 2, and 4 right away no problem =]

About 3, I thought of having a producer of `ContextPropagators` because this would contain all propagators, so someone can override the bean and produce something like `ContextPropagators.create(TextMapPropagator.composite(W3CTraceContextPropagator.getInstance(), W3CBaggagePropagator.getInstance()));` . But it would be just a beans, also thought about having a bean lookup of `TextMapPropagator` and in the recorder join then in a  `TextMapPropagator.composite` that would do mostly the same thing in the end.

About the Resource, I kind thought about the same way. If someone wants multiple resources they can use `.merge` within the bean producer, or we can have multiple beans and let the recorder join them, the only thing would be the order of precedence, since when merging a desired property can be left out since it they are mostly a List and before adding it checks if the key is already present.

What do you think? Do you already have some proposal for those?

Happy to help =]

**Comment by kenfinnigan** [comment]
Thanks @luneo7, if you could do 1, 2, 4, that would be greatly appreciated!

As for 3, I hadn't done anything beyond some thinking of the various options.

For propagators, I was thinking of providing a config option to define the propagators you want to be installed, similar to what [OpenTelemetry Java AutoConfigure](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure) does, but doing so in a way that doesn't require the extension to depend on the artifact which includes them. So it would need some mapping of "type name" to "class name string" so we can see if the class exists before we install a single or composite propagator.

Though we could have user applications create composite propagators, if we can keep it simple to do the developer experience should be improved. Do you agree?

Resources are certainly more tricky, but there are also some aspects of the AutoConfigure extension providing default resources for environment, process, etc which could be interesting to provide without a developer having to construct a Resource with all the necessary bits

**Comment by luneo7** [comment]
One thing about AutoConfigure is that they created one SPI, and they are using `compileOnly` dependencies so they can use some classes there without doing reflection. If we were going to do a class mapping, we would need more than the class name, because some of them have different methods to get the instance, and we would have to do reflection to instantiate the propagators, is that OK (like `B3Propagator.injectingSingleHeader()` and `B3Propagator.injectingMultiHeaders()`?

**Comment by kenfinnigan** [comment]
Yeah, I saw how they did that, and I'd certainly like to avoid the need for reflection.

I've asked a [question](https://github.com/open-telemetry/opentelemetry-java/discussions/3284) on whether there's a reason for some differences between the propagators, as the AWS propagator implements a provider with an SPI method for retrieving the instance.

If the other propagators had providers as well it would make implementation easier.

Otherwise, I don't have a good answer for how to solve it right now.

**Comment by luneo7** [comment]
@kenfinnigan Updated the PR, now it has only the OTLP exporter

**Comment by luneo7** [comment]
@kenfinnigan Created another PR to address the config params: https://github.com/quarkusio/quarkus/pull/17703

File: extensions/opentelemetry/opentelemetry-exporter-otlp/deployment/pom.xml [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-opentelemetry-exporter-otlp-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-opentelemetry-exporter-otlp-deployment</artifactId>
+    <name>Quarkus - OpenTelemetry Exporter - Otlp - Deployment</name>
```
Comment (kenfinnigan) [inline extensions/opentelemetry/opentelemetry-exporter-otlp/deployment/pom.xml:14]: It's minor, but should "Otlp" be all capitalized? I usually see it in all capitals.

File: extensions/opentelemetry/opentelemetry-exporter-otlp/deployment/src/test/java/io/quarkus/opentelemetry/exporter/otlp/deployment/OtlpExporterDisabledTest.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class))
+            .overrideConfigKey("quarkus.opentelemetry.tracer.exporter.otlp.enabled", "false");
+
+    @Inject
+    OpenTelemetry openTelemetry;
+
+    @Test
```
Comment (kenfinnigan) [inline extensions/opentelemetry/opentelemetry-exporter-otlp/deployment/src/test/java/io/quarkus/opentelemetry/exporter/otlp/deployment/OtlpExporterDisabledTest.java:24]: Should this test also include the same checks the Jaeger Exporter has verifying the `LateBoundBatchSpanProcessor` isn't resolvable?

File: extensions/opentelemetry/opentelemetry-exporter-otlp/runtime/src/main/resources/META-INF/quarkus-extension.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+name: "OpenTelemetry exporter: Otlp"
```
Comment (kenfinnigan) [inline extensions/opentelemetry/opentelemetry-exporter-otlp/runtime/src/main/resources/META-INF/quarkus-extension.yaml:1]: Capitalize "Otlp"?

**Review by kenfinnigan — CHANGES_REQUESTED** [review]
Thanks @luneo7, looks really good.

Had a couple of minor comments, but the big question is whether we should have a section in the OpenTelemetry Guide talking about how to use the OTLP Exporter instead of the Jaeger one. What do you think?

**Comment by luneo7** [comment]
I think that having the OTLP Exporter in the guide would be preferable since it is the open standard protocol and the collector can export it later to whatever people want. And right now the difference between the two exporters in terms of configuration are minimal, and we can even add a note that switching to the Jaeger protocol would be as easy as using the Jaeger exporter import instead of the OTLP.

**Comment by kenfinnigan** [comment]
Sounds good, could you add that with this PR? Thanks

**Comment by luneo7** [comment]
Yup, saw that we already have the OTEL guide (https://github.com/quarkusio/quarkus/pull/17639), will tweak it =]

File: extensions/opentelemetry/opentelemetry-exporter-otlp/runtime/pom.xml [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-grpc-common</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>io.opentelemetry</groupId>
+            <artifactId>opentelemetry-exporter-otlp</artifactId>
+            <exclusions>
```
Comment (kenfinnigan) [inline extensions/opentelemetry/opentelemetry-exporter-otlp/runtime/pom.xml:39]: One other thing I just thought of.

For the moment, we should probably only depend on `opentelemetry-exporter-otlp-trace` as we're not supporting exports for anything other than tracing right now.

When we add other signal exporting in the future we can switch back to the full dependency.

Comment (luneo7) [reply extensions/opentelemetry/opentelemetry-exporter-otlp/runtime/pom.xml:39]: Yup, we can make this just the `oltp-trace`, but today the `exporter-otlp` only depends on the `oltp-trace`. But we can change it to make sure if the `all` pom changes we are just with the trace dependencies.

image

Comment (kenfinnigan) [reply extensions/opentelemetry/opentelemetry-exporter-otlp/runtime/pom.xml:39]: Right, probably best to explicitly depend on just the `trace` artifact as I expect at some point soon some of that will change, and don't want to bring it in until we're ready to handle it

**Comment by luneo7** [comment]
@kenfinnigan updated the docs and also created a PR for the quick-start (https://github.com/quarkusio/quarkus-quickstarts/pull/877)

**Review by kenfinnigan — APPROVED** [review]

**Comment by kenfinnigan** [comment]
Thanks @luneo7!

Really appreciate the contribution.

One thing I hadn't thought of, until your doc and quickstart changes, is that we might want to move the jaeger exporter into the Quarkiverse as opposed to being a main extension. What do you think?

**Comment by luneo7** [comment]
I think it makes sense to move the Jaeger exporter to the Quarkiverse, since we already have everything to make it work in the main extensions and nothing "core" depends on it.

**Pull request merged**